First implementation of new SI API

TedHartMS opened the pull request.

TedHartMS committed:
First implementation of new SI API

badrishc commented on cs/benchmark/SecondaryIndexes.cs:
```diff
@@ -0,0 +1,33 @@
+
+        public void Delete(ref Key key) { }
+
+        public void Insert(ref Key key) { }
```
There is a variant here that takes in the address as well. In that case, we need to update the index whenever addresses changes, even if the key does not change. Like an Upsert call here in addition to Insert.

badrishc commented on cs/benchmark/SecondaryIndexes.cs:
```diff
@@ -0,0 +1,33 @@
+
+        public void Insert(ref Value value, long recordId) { }
+
+        public void Upsert(ref Value value, long recordId, bool isMutable) { }
```
What does isMutable mean here in the API?

badrishc commented on cs/src/core/Allocator/GenericAllocator.cs:
```diff
@@ -922,7 +922,7 @@ protected override bool RetrievedFullRecord(byte* record, ref AsyncIOContext<Key
         public override bool KeyHasObjects()
         {
-            return SerializerSettings.keySerializer != null;
+            return SerializerSettings?.keySerializer != null;
```
Check where the nullref occurred and try to fix the caller spot.

badrishc commented on cs/src/core/ClientSession/AdvancedClientSession.cs:
```diff
@@ -737,15 +737,54 @@ public void ConcurrentReader(ref Key key, ref Input input, ref Value value, ref
             public bool ConcurrentWriter(ref Key key, ref Value src, ref Value dst, long address)
             {
-                return _clientSession.functions.ConcurrentWriter(ref key, ref src, ref dst, address);
+                if (!_clientSession.fht.SupportsMutableIndexes)
```
* Make `SupportsMutableIndexes` a field of this struct and see if it helps.
* Other thing to try is to make this struct itself readonly -- will that help.
* Moving else path to non-inlined extra heavy-method might help the fast path?

badrishc commented on cs/src/core/Index/FASTER/FASTER.cs:
```diff
@@ -591,6 +602,18 @@ public async ValueTask CompleteCheckpointAsync(CancellationToken token = default
+                        if (this.SecondaryIndexBroker.MutableValueIndexCount > 0)
+                            this.SecondaryIndexBroker.Insert(ref value, pcontext.logicalAddress);
+                    }
+                    recordInfo.Unlock();
```
where is the .Lock?

badrishc commented on cs/src/core/Index/FASTER/FASTER.cs:
```diff
@@ -618,6 +641,18 @@ public async ValueTask CompleteCheckpointAsync(CancellationToken token = default
+                        if (this.SecondaryIndexBroker.MutableValueIndexCount > 0)
+                            this.SecondaryIndexBroker.Insert(ref this.hlog.GetValue(this.hlog.GetPhysicalAddress(pcontext.logicalAddress)), pcontext.logicalAddress);
+                    }
+                    recordInfo.Unlock();
```
where is the .Lock()?

badrishc commented on cs/src/core/ClientSession/ClientSession.cs:
```diff
@@ -747,13 +747,53 @@ public void ConcurrentReader(ref Key key, ref Input input, ref Value value, ref
+                if (!_clientSession.fht.SupportsMutableIndexes)
+                    return _clientSession.functions.ConcurrentWriter(ref key, ref src, ref dst);
+
+                ref RecordInfo recordInfo = ref _clientSession.fht.RecordAccessor.SpinLockRecordInfo(address);
```
It's a bit worrisome that locking is all over the place and using spin locks.

Scenarios to consider:
* What if the value were 8 byte longs and so we didn't need a lock at all?
* Or the user is okay with broken writes (value consists of individual 8-byte values that don't all need to be consistently read at ones)
* What if we were only having key indexes, so we never need to take a lock at all (since keys are immutable)?
* What if the user had space in key/value to have a better implementation of lock.

Would it be better if we made this a duty of (advanced) functions - capable of being overridden by specific functions. like `functions.LockRecord(ref RecordInfo, ref Key, ref Value)`.

badrishc commented on cs/src/core/ClientSession/ClientSession.cs:
```diff
@@ -747,13 +747,53 @@ public void ConcurrentReader(ref Key key, ref Input input, ref Value value, ref
+                {
+                    if (!recordInfo.Tombstone && _clientSession.functions.ConcurrentWriter(ref key, ref src, ref dst))
+                    {
+                        // KeyIndexes do not need notification of in-place updates because the key does not change.
```
The lock seems wasteful to take and release if we didnt need to notify SI.

badrishc commented on cs/src/core/ClientSession/ClientSession.cs:
```diff
@@ -747,13 +747,53 @@ public void ConcurrentReader(ref Key key, ref Input input, ref Value value, ref
+                if (!_clientSession.fht.SupportsMutableIndexes)
+                    return false;
+
+                ref RecordInfo recordInfo = ref _clientSession.fht.RecordAccessor.SpinLockRecordInfo(address);
```
Again, we are taking a lock even when we do not need to talk to SI.

badrishc commented on cs/src/core/SecondaryIndex/SecondaryIndexBroker.cs:
```diff
@@ -0,0 +1,151 @@
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Insert(ref TKVKey key)
+        {
+            foreach (var keyIndex in mutableKeyIndexes)
```
Is this iteration thread-safe with adding indexes?

badrishc commented on cs/src/core/ClientSession/ClientSession.cs:
```diff
@@ -747,13 +747,53 @@ public void ConcurrentReader(ref Key key, ref Input input, ref Value value, ref
+                    if (!recordInfo.Tombstone && _clientSession.functions.ConcurrentWriter(ref key, ref src, ref dst))
+                    {
+                        // KeyIndexes do not need notification of in-place updates because the key does not change.
+                        if (_clientSession.fht.SecondaryIndexBroker.MutableValueIndexCount > 0)
```
Keys in FASTER are immutable once added to the log, so we do not need a lock when only accessing keys.

TedHartMS committed:
Add I(Advanced)Functions.Lock/Unlock and minimize SI-related locking

TedHartMS committed:
some experimental stuff to test

TedHartMS committed:
Experiment with SupportsLocks

TedHartMS committed:
Change to new locking scheme; update test Functions to derive from FunctionsBase

badrishc commented on cs/src/core/VarLen/SpanByteFunctions.cs:
```diff
@@ -10,19 +10,11 @@ namespace FASTER.core
-        {
-            this.locking = locking;
-        }
+        public SpanByteFunctions(bool locking = false) : base(locking) { }
```
Why are we still doing locking inside ConcurrentWriter below? Shouldn't it be refactored into Lock() and Unlock() overrides.

badrishc commented on cs/src/core/VarLen/MemoryFunctions.cs:
```diff
@@ -11,17 +11,15 @@ public class MemoryFunctions<Key, T, Context> : FunctionsBase<Key, Memory<T>, Me
         /// <param name="memoryPool"></param>
         /// <param name="locking">Whether we lock values before concurrent operations (implemented using a spin lock on length header bit)</param>
-        public MemoryFunctions(MemoryPool<T> memoryPool = default, bool locking = false)
+        public MemoryFunctions(MemoryPool<T> memoryPool = default, bool locking = false) : base(locking)
```
Why are we still doing locking inside ConcurrentWriter below? Shouldn't it be refactored into Lock() and Unlock() overrides.

badrishc commented on cs/test/ObjectTestTypes.cs:
```diff
@@ -328,13 +281,13 @@ public class MyLargeOutput
     }
 
-    public class MyLargeFunctions : IFunctions<MyKey, MyLargeValue, MyInput, MyLargeOutput, Empty>
+    public class MyLargeFunctions : FunctionsBase<MyKey, MyLargeValue, MyInput, MyLargeOutput, Empty>
```
It would be good practice to `seal` all overridden user functions.

badrishc commented on cs/src/core/SecondaryIndex/SecondaryIndexBroker.cs:
```diff
@@ -0,0 +1,151 @@
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Delete(long recordId)
+        {
+            foreach (var valueIndex in mutableValueIndexes)
```
if `foreach` as efficient in assembly code as a normal for loop?

badrishc commented on cs/src/core/Index/Common/Contexts.cs:
```diff
@@ -12,12 +12,20 @@
         RMW,
+        /// <summary>Upsert operation</summary>
         UPSERT,
+        /// <summary>Insert operation (either Upsert or RMW resulted in a new record)</summary>
```
Is INSERT also used in internal use of `OperationType`. Why not keep a separate enum for the user facing API compared to internal?

badrishc commented on cs/src/core/ClientSession/AdvancedClientSession.cs:
```diff
@@ -763,14 +836,35 @@ public int GetLength(ref Value t, ref Input input)
             {
-                return _clientSession.functions.InPlaceUpdater(ref key, ref input, ref value, address);
+                long context = 0;
+                this.Lock(ref recordInfo, ref key, ref value, OperationType.RMW, ref context);
```
It seems overkill to send OperationType to Lock. All we need to know here is whether this is a reader lock or writer lock. That is why I suggested either a new enum `OpType` or just a bool.

badrishc replied:
Maybe call it LockType.Shared and Exclusive?

TedHartMS committed:
Additional locking work:
- enable locking in YCSB benchmark
- re-allow threadAffinitized sessions with indexes (still need to add some SI calls in pending handlers)
- rename to .SupportsLocking
- create LockType enum rather than reusing OperationType
- make tests for context.IsNewRecord more efficient in ContextUpsert et al.

TedHartMS committed:
Update YCSB benchmark to separate locking and indexing parameters and add an iteration count; refactor Memory and Span Functions to use the new locking scheme

badrishc committed:
updates

TedHartMS committed:
Improvments to multi-iteration runs

TedHartMS committed:
Remove FKV.SupportsMutableIndexes; change benchmark to resuse loaded files, validate params, generate parsable output

TedHartMS committed:
Benchmark: remove interface and make store, device, keys readonly

TedHartMS committed:
distinct delimiter on last and trimmed stats summary

TedHartMS committed:
More YCSB benchmark updates:
- Force synthetic data for kUseSmallData
- Add the Zipf generator for synthetic data
- Improve output formatting so it is more easily parsable and has all needed info

badrishc committed:
Update benchmark to start in unison.

TedHartMS committed:
Refactor YCSB to clean up program.cs and duplicate LoadData and improve output

TedHartMS committed:
Fix UpdateSIForIPU return; fix latchDest in InternalRMW; tweak addresses StatsLine

TedHartMS committed:
turn off smalldata for Release

TedHartMS committed:
YCSB changes:
- Move backup/restore to subdirs dependent upon distribution, ycsb vs. synthetic data, and sizes
- Reduce -k (backup) parameter from bitmask to bool
- Move remaining constants to YcsbConstants

TedHartMS committed:
YCSB: change --backup to --recover and remove BackupMode; add --runsec

TedHartMS committed:
a few minor tweaks made while doing PRs

TedHartMS committed:
Add Powershell scripts to run the benchmark performance suite and compare results

TedHartMS committed:
Merge upstream master

TedHartMS committed:
updates to benchmark scripts

TedHartMS committed:
Merge remote-tracking branch 'origin/master' into SecondaryIndex

TedHartMS committed:
Add SubsetIndexSessionBroker; add SimpleIndex tests

TedHartMS committed:
Implement ReadOnlyObserver for SI; add Immutable and Mixed index test scenarios

TedHartMS committed:
Support multiple ReadOnly and Eviction Observers

TedHartMS committed:
Merge upstream master

TedHartMS committed:
more files from merge

TedHartMS committed:
Merge remote-tracking branch 'origin/master' into SecondaryIndex

TedHartMS committed:
Add RecordId struct; move QueryRecord into SI from HVI; remove recordId from Key Index and split out the Key vs. Value index test base classes; make IHeapContainer inherit from IDisposable; add SI calls to UpsertAsync

TedHartMS committed:
Add QueryRecord.RecordId

TedHartMS committed:
RecordId changes from HVI

TedHartMS committed:
QueryRecord changes from HVI

TedHartMS committed:
Fixes to multi-observer cleanup; fix recordInfo setting in ReadAsyncInternal; Fix QueryRecord.Dispose(); make RecordId.IsDefault() a method so it's not serialized

TedHartMS committed:
merge upstream master

TedHartMS committed:
Modify Secondary Index interfaces:
- Add RIDs to Key indexes
- Add TKVKEey to Value indexes
- Add Checkpoint/Restore API
- Add Truncation API

TedHartMS committed:
Merge upstream master

TedHartMS committed:
Merge upstream master; backport from HVI branch

TedHartMS committed:
Merge remote-tracking branch 'origin/master' into SecondaryIndex

TedHartMS committed:
allow deltaFileDevice == null

TedHartMS committed:
Additional changes for C#9

TedHartMS committed:
Backport a few changes from HVI in preparation to merge master

TedHartMS committed:
Merge upstream master

TedHartMS committed:
merge upstream master

TedHartMS committed:
add missing OnRecovery to InternalRecoverAsync

TedHartMS committed:
Add default values for scanDelta, recoverTo args

TedHartMS committed:
backport mostly doc-related changes from HVI
